### PR TITLE
Expose Test and TestOutput events as build operations

### PR DESCRIPTION
**Pull request by breskeby**
### Context
This PR changes Gradle test execution to expose Test and TestOutput events as Build Operations. 

This allows build operation consumers like the build scan plugin to listen for all test events in an explicit or implicit composite builds and observing historic operations fired before consumer was registered.

A crucible review has been created for convenience at https://code-review.gradle.org/cru/REVIEW-6569#CFR-125214

### CODE REVIEW

**Comment by eskatos** [comment]
@breskeby, could you please remove spurious changes from your branch so the review is easier? (broken javadoc formatting, blank lines here and there etc..)

**Comment by breskeby** [comment]
@eskatos fixed formatting

**Comment by eskatos** [comment]
[CI Branch Build](https://builds.gradle.org/viewQueued.html?itemId=8975833)

File: subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+import org.gradle.api.tasks.testing.ExecuteTestBuildOperationType
+import org.gradle.api.tasks.testing.TestOutputBuildOperationType
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.integtests.fixtures.TestResources
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.junit.Rule
+
+class TestListenerBuildOperationAdapterIntegrationTest extends AbstractIntegrationSpec {
```
Comment (ldaley) [inline subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy]: Ideally, we would be reusing our existing strategies for asserting on our understanding of the test execution, which is based around `TestExecutionResult`. We already have a bunch of tests that assert how we represent different quirky test arrangements in our model, of which these new operations are now effectively part. 

In practice though, this may not be worth it. It would be some work to have TestExecutionResult transparently assert on the event stream, if it is at all possible without drastic changes.

My overarching concern here is that functional tests for this new feature are patchy. We either need to make that not the case, our base our QA on implicit comprehensive coverage (e.g. we assume that correct HTML test result reports translates into correct emission of operation notifications).

Comment (ldaley) [inline subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy]: This class is poorly named as it leaks unnecessary implementation specifics. Should be `TestExecutionBuildOperationsIntegrationTest` or something similar.

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy [hunk 2]
```diff
@@ -281,5 +281,6 @@ class TestTest extends AbstractConventionTaskTest {
         test.setBinResultsDir(binResultsDir)
         test.getReports().getHtml().setDestination(reportDir)
         test.setClasspath(classpathMock)
+        test.setTestListenerBuildOperationAdapter(Mock(TestListenerBuildOperationAdapter))
```
Comment (ldaley) [inline subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy]: Why is this necessary?

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTaskSpec.groovy [hunk 3]
```diff
@@ -29,12 +29,14 @@ class TestTaskSpec extends AbstractProjectBuilderSpec {
...
     def suiteDescriptor = Mock(TestDescriptorInternal)
     def testDescriptor = Mock(TestDescriptorInternal)
+    def testBuildOperationAdapter = Mock(TestListenerBuildOperationAdapter)
 
     private WorkerLeaseRegistry.WorkerLeaseCompletion completion
     private Test task
 
     def setup() {
         task = TestUtil.create(temporaryFolder).task(Test, [testExecuter: testExecuter, testFramework: testFramework])
+        task.testListenerBuildOperationAdapter = testBuildOperationAdapter
```
Comment (ldaley) [inline subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTaskSpec.groovy]: Why is this necessary?

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    public TestListenerBuildOperationAdapter(BuildOperationState testRootBuildOperation, BuildOperationListener listener, BuildOperationIdFactory buildOperationIdFactory, Clock clock) {
+        this.testRootBuildOperation = testRootBuildOperation;
+        this.listener = listener;
+        this.buildOperationIdFactory = buildOperationIdFactory;
+        this.clock = clock;
+    }
+
+    @Override
+    public void started(final TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
+        if (testDescriptor.getParent() != null) {
```
Comment (ldaley) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java]: Why are we deviating from the model already used elsewhere where we wrap a “suite” around the whole tree?

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            this.event = event;
+        }
+
+        @Override
+        public TestOutputEvent getOutput() {
+            return event;
+        }
+    }
+
+    private class DefaultTestBuildOperationDetails implements ExecuteTestBuildOperationType.Details {
```
Comment (ldaley) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java]: Should be static.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            .operationType(BuildOperationCategory.TASK)
+            .details(details)
+            .build(newOperationIdentifier(), parentId);
+    }
+
+    private OperationIdentifier newOperationIdentifier() {
+        return new OperationIdentifier(buildOperationIdFactory.nextId());
+    }
+
+    private class DefaultTestOutputBuildOperationResult implements TestOutputBuildOperationType.Result {
```
Comment (ldaley) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java]: Should be static.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/BuildOperationTestResult.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.testing;
+
+/**
+ * Result describing Test Build Operation Result
+ */
+public class BuildOperationTestResult implements ExecuteTestBuildOperationType.Result {
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/BuildOperationTestResult.java]: Why is this not inlined into the adapter like the others?

**Comment by ldaley** [comment]
This will need review from someone familiar with the test status logging. @breskeby please arrange that.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        public TestDescriptor getTestDescriptor() {
+            return testDescriptor;
+        }
+
+        public long getStartTime() {
+            return startTime;
+        }
+    }
+
+    private class TestBuildOperation {
```
Comment (ldaley) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java]: This could be named better. Perhaps `InProgressExecuteTestBuildOperation`.

File: subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy (hunk 1 above)
Comment (breskeby) [reply subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy]: one alternative would be to generate the trace file and use that to gather the results in a TextExecutionResult. on the plus side it would be way easier to get TestExecutionResult transparently with that. The downside is that it comes with quite a cost.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 9]
```diff
@@ -431,6 +450,14 @@ public void executeTests() {
 
         TestExecuter testExecuter = createTestExecuter();
 
+        if (testListenerBuildOperationAdapter == null) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: Currently core doesn't know anything about gradle test infrastructure. moving this into `BuildSessionScopeServices` would need to change that.

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTaskSpec.groovy (hunk 3 above)
Comment (breskeby) [reply subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTaskSpec.groovy]: not mocking would result in a `java.lang.IllegalStateException: No operation is currently running.`
thrown by the `DefaultBuildOperationExecutor` when calling `getBuildOperationExecutor().getCurrentOperation()`.

I guess the alternative would be to change `ProjectBuilderImpl` which I thought is too invasive.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: `testing-base` could have its own `PluginServiceRegistry`

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy (hunk 2 above)
Comment (breskeby) [reply subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy]: see above

File: subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy (hunk 1 above)
Comment (breskeby) [reply subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestListenerBuildOperationAdapterIntegrationTest.groovy]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: I've implemented the TestListenerBuildOperationAdapter as BuildSessionScoped service but kept it as explicit `TestListenerBuildOperationAdapter`. I think since it's misleading as some some TaskListenerInternal implementation currently require manual instantiation (e.g. `TestWorkerProgressListener`. Therefore it's cleaner to me to use explicit `TestListenerBuildOperationAdapter` here

**Comment by breskeby** [comment]
@oehme / @adammurdoch  I think you both have most knowledge regarding test status logging. would you mind having a look on this PR. I've adding you as requested reviewers

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+import org.gradle.internal.progress.OperationStartEvent;
+import org.gradle.internal.time.Clock;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Emitting build operations for tests
+ */
+public class TestListenerBuildOperationAdapter implements TestListenerInternal {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestListenerBuildOperationAdapter.java]: moved this into an internal package and made it package private.

File: subprojects/distributions/src/changes/accepted-public-api-changes.json [hunk 11]
```diff
@@ -165,6 +165,51 @@
             "type": "org.gradle.api.tasks.testing.AbstractTestTask",
             "member": "Constructor org.gradle.api.tasks.testing.AbstractTestTask()",
             "acceptation": "Extracted super class out of Test which was already stable"
-        }
+        },
+        {
+            "type": "org.gradle.api.tasks.testing.ExecuteTestBuildOperationType",
+            "member": "Class org.gradle.api.tasks.testing.ExecuteTestBuildOperationType",
+            "acceptation": "Build Operation typs are not annotated with incubating"
```
Comment (eskatos) [inline subprojects/distributions/src/changes/accepted-public-api-changes.json]: :x: typo `typs` -> `types`, same in others

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+import org.gradle.internal.time.Clock;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Emitting build operations for tests
+ */
+public class TestListenerBuildOperationAdapter implements TestListenerInternal {
+    public static final TestOutputBuildOperationType.Details TEST_OUTPUT_DETAILS = new TestOutputBuildOperationType.Details() {
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: :x: can be `private`

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        public TestDescriptor getTestDescriptor() {
+            return testDescriptor;
+        }
+
+        public long getStartTime() {
+            return startTime;
+        }
+    }
+
+    private class BuildOperationTestResult implements ExecuteTestBuildOperationType.Result {
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: :x: can be `static`

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+            this.result = testResult;
+        }
+
+        @Override
+        public TestResult getResult() {
+            return result;
+        }
+    }
+
+    private class InProgressExecuteTestBuildOperation {
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: :x: can be `static`

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 15]
```diff
@@ -78,11 +78,7 @@
 /**
  * Abstract class for all test task.
  *
- * <ul>
- *     <li>Support for test listeners</li>
- *     <li>Support for reporting</li>
- *     <li>Support for report linking in the console output</li>
- * </ul>
+ * <ul> <li>Support for test listeners</li> <li>Support for reporting</li> <li>Support for report linking in the console output</li> </ul>
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: broken javadoc formatting

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 16]
```diff
@@ -167,9 +170,14 @@ protected ListenerManager getListenerManager() {
         return testListenerInternalBroadcaster;
     }
 
-    /**
-     * ATM. for testing only
-     */
+    @Internal
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: this annotation is for getters only and should be removed, see its javadoc

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 17]
```diff
@@ -167,9 +170,14 @@ protected ListenerManager getListenerManager() {
...
+    @VisibleForTesting
+    void setTestListenerInternal(TestListenerInternal testListenerInternal) {
+        this.testListenerBuildOperationAdapter = testListenerInternal;
+    }
+
+    @Internal
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: this annotation is for getters only and should be removed, see its javadoc

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 18]
```diff
@@ -274,20 +282,11 @@ private TestExceptionFormatter getExceptionFormatter(TestLogging testLogging) {
...
      *
-     * test {
-     *    onOutput { descriptor, event -&gt;
-     *        if (event.destination == TestOutputEvent.Destination.StdErr) {
-     *            logger.error("Test: " + descriptor + ", error: " + event.message)
-     *        }
-     *    }
-     * }
-     * </pre>
+     * test { onOutput { descriptor, event -&gt; if (event.destination == TestOutputEvent.Destination.StdErr) { logger.error("Test: " + descriptor + ", error: " + event.message) } } } </pre>
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: broken javadoc formatting

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 19]
```diff
@@ -339,13 +337,9 @@ public void afterTest(Closure closure) {
...
      *
-     * <pre class='autoTested'>
-     * apply plugin: 'java'
+     * <pre class='autoTested'> apply plugin: 'java'
      *
-     * test.testLogging {
-     *     exceptionFormat "full"
-     * }
-     * </pre>
+     * test.testLogging { exceptionFormat "full" } </pre>
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: broken javadoc formatting

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 20]
```diff
@@ -360,14 +354,9 @@ public TestLoggingContainer getTestLogging() {
...
-     * <pre class='autoTested'>
-     * apply plugin: 'java'
+     * <pre class='autoTested'> apply plugin: 'java'
      *
-     * // makes the standard streams (err and out) visible at console when running tests
-     * test.testLogging {
-     *    showStandardStreams = true
-     * }
-     * </pre>
+     * // makes the standard streams (err and out) visible at console when running tests test.testLogging { showStandardStreams = true } </pre>
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: broken javadoc formatting

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 21]
```diff
@@ -378,14 +367,9 @@ public void testLogging(Closure closure) {
...
```
Comment (eskatos) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: :x: broken javadoc formatting

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingBuildOperationListener.java [hunk 22]
```diff
@@ -40,12 +48,21 @@
 
     @Override
     public void started(BuildOperationDescriptor buildOperation, OperationStartEvent startEvent) {
-        eventConsumer.dispatch(new DefaultOperationStartedProgressEvent(startEvent.getStartTime(), toBuildOperationDescriptor(buildOperation)));
+        if (!isTestRelatedBuildOperationEvent(buildOperation.getDetails())) {
+            eventConsumer.dispatch(new DefaultOperationStartedProgressEvent(startEvent.getStartTime(), toBuildOperationDescriptor(buildOperation)));
+        }
+    }
+
+    private boolean isTestRelatedBuildOperationEvent(Object details) {
```
Comment (eskatos) [inline subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingBuildOperationListener.java]: :o: I'd move this method below both its usages, for readability

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/ExecuteTestBuildOperationType.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+     * Details of a running Test
+     *
+     * @since 4.4
+     * */
+    @UsedByScanPlugin
+    public interface Details {
+        /**
+         * the descriptor of the executed test.
+         * */
+        TestDescriptor getTestDescriptor();
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/ExecuteTestBuildOperationType.java]: ❌ This would make `TestDescriptor` part of the protocol between the build tool and the scan plugin. I think this is contrary to the "just what we need" approach we used in other places.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java [hunk 24]
```diff
@@ -167,9 +170,14 @@ protected ListenerManager getListenerManager() {
         return testListenerInternalBroadcaster;
     }
 
-    /**
-     * ATM. for testing only
-     */
+    @Internal
+    @VisibleForTesting
+    void setTestListenerInternal(TestListenerInternal testListenerInternal) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: Instead of adding a setter, you could use subclassing in the unit test. That would make the production code simpler to read.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Override
+    public void output(final TestDescriptorInternal testDescriptor, final TestOutputEvent event) {
+        InProgressExecuteTestBuildOperation runningOp = runningTests.get(testDescriptor);
+        BuildOperationDescriptor.Builder outputDescription = BuildOperationDescriptor.displayName(event.getDestination().name())
+            .operationType(BuildOperationCategory.TASK)
+            .details(TEST_OUTPUT_DETAILS);
+        BuildOperationDescriptor outputBuildOperationDescriptor = outputDescription.build(newOperationIdentifier(), runningOp.descriptor.getId());
+        long currentTime = clock.getCurrentTime();
+        listener.started(outputBuildOperationDescriptor, new OperationStartEvent(currentTime));
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: ❌ This doesn't really make sense to me - Starting and finishing a build operation for each piece of output a test creates. That would clutter the trace output pretty quickly. Wouldn't it be better to aggregate the output in the test finished event?

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+        listener.finished(outputBuildOperationDescriptor, new OperationFinishEvent(currentTime, currentTime, null, new DefaultTestOutputBuildOperationResult(event)));
+    }
+
+    private BuildOperationDescriptor createTestBuildOperationDescriptor(TestDescriptor testDescriptor, TestStartEvent testStartEvent) {
+        DefaultTestBuildOperationDetails details = new DefaultTestBuildOperationDetails(testDescriptor, testStartEvent.getStartTime());
+        InProgressExecuteTestBuildOperation parentOperation = runningTests.get(testDescriptor.getParent());
+        Object parentId = parentOperation == null ? buildOperationExecutor.getCurrentOperation().getId() : parentOperation.descriptor.getId();
+        return BuildOperationDescriptor.displayName(testDescriptor.getName())
+            .operationType(BuildOperationCategory.TASK)
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: ❌ This is not a task, so it seems this is the wrong category.

File: subprojects/distributions/src/changes/accepted-public-api-changes.json [hunk 27]
```diff
@@ -165,6 +165,51 @@
             "type": "org.gradle.api.tasks.testing.AbstractTestTask",
             "member": "Constructor org.gradle.api.tasks.testing.AbstractTestTask()",
             "acceptation": "Extracted super class out of Test which was already stable"
-        }
+        },
+        {
+            "type": "org.gradle.api.tasks.testing.ExecuteTestBuildOperationType",
```
Comment (oehme) [inline subprojects/distributions/src/changes/accepted-public-api-changes.json]: ❌ These shouldn't be part of the public API, they are only for the scan plugin.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestIgnoringBuildOperationListener.java [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+package org.gradle.tooling.internal.provider.runner;
+
+import org.gradle.api.tasks.testing.ExecuteTestBuildOperationType;
+import org.gradle.api.tasks.testing.TestOutputBuildOperationType;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.internal.progress.BuildOperationListener;
+import org.gradle.internal.progress.OperationFinishEvent;
+import org.gradle.internal.progress.OperationStartEvent;
+
+class TestIgnoringBuildOperationListener implements BuildOperationListener {
```
Comment (oehme) [inline subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestIgnoringBuildOperationListener.java]: ❌ Missing class documentation.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java (hunk 25 above)
Comment (ldaley) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: We are currently missing the concept of a single point in time event in build operation land, like the progress events. When we move to having the plugin observe logging output via operations, we'll have the same problem.

Forcing Gradle to always buffer all of the test output in memory until the test finishes seems like a big sacrifice (fwiw, the scan plugin is already doing this but its a different thing doing it there in the plugin).

Alternatively, we enrich the operation event and notification APIs to support point-in-time events like the existing progress model. I'm leaning towards that.

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/ExecuteTestBuildOperationType.java (hunk 23 above)
Comment (ldaley) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/ExecuteTestBuildOperationType.java]: This is already stable public API. I don't see why we want to create more objects or another data model.

**Comment by ldaley** [comment]
@eskatos @oehme I'm unsure how much to invest in the functional testing here. Right now, it's pretty thin and effectively duplicates other tests in terms of scenario under test. Ideally, we'd only test each scenario (e.g. particular JUnit setup once) and in that test assert that we interpreted and expressed the results in all the ways that we do (e.g. console output, HTML report, test listener notifications, build operation notifications). 

With the work happening to use more of this infrastructure for other kinds of tests (e.g. xcode test), this seems like it's going to be more important. 

How do you want to handle this?

**Comment by eskatos** [comment]
Yep, the assertions on emitted build operations could be added to existing functional tests instead of creating new tests. This would limit the proliferation of tests thus preventing added test execution time.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestIgnoringBuildOperationListener.java (hunk 28 above)
Comment (breskeby) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestIgnoringBuildOperationListener.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 21 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 20 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 19 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 18 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 17 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: removed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 16 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: removed

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 15 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java (hunk 14 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java (hunk 13 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java (hunk 12 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: fixed

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java (hunk 26 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/TestListenerBuildOperationAdapter.java]: I thought task category would fit here as its running in the context of a task. removed the category

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 15 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: done

File: subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java (hunk 24 above)
Comment (breskeby) [reply subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java]: done.

File: subprojects/distributions/src/changes/accepted-public-api-changes.json (hunk 11 above)
Comment (breskeby) [reply subprojects/distributions/src/changes/accepted-public-api-changes.json]: as we made those build operation types internal I removed them from the binary check json file completely

File: subprojects/distributions/src/changes/accepted-public-api-changes.json (hunk 27 above)
Comment (breskeby) [reply subprojects/distributions/src/changes/accepted-public-api-changes.json]: moved to internal package

**Comment by breskeby** [comment]
@eskatos @ldaley I think it makes sense to move the build operation based test assertions into the general test result fixtures. But I would propose to do that in a separate PR as it requires some more work and is easier to handle in a separate story. We can create a follow up issue on this. WDYT?

**Comment by breskeby** [comment]
merge conflicts have been resolved

**Comment by breskeby** [comment]
CI has passed at https://builds.gradle.org/viewLog.html?buildTypeId=Gradle_Check_Stage_BranchBuildAccept_Trigger&buildId=9044642&branch_Gradle_Check_Stage_BranchBuildAccept_Trigger=breskeby/test-build-operations

File: subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestExecutionBuildOperationsIntegrationTest.groovy [hunk 29]
```diff
@@ -0,0 +1,191 @@
...
+import org.gradle.api.internal.tasks.testing.operations.TestOutputBuildOperationType
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.integtests.fixtures.TestResources
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.junit.Rule
+
+import static org.gradle.util.TextUtil.normaliseLineSeparators
+
+class TestExecutionBuildOperationsIntegrationTest extends AbstractIntegrationSpec {
```
Comment (big-guy) [inline subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestExecutionBuildOperationsIntegrationTest.groovy]: We also have XCTests -- do we need to a separate test for build operations for that as well?

Maybe we could move the first two "emits..." tests into `AbstractTestFrameworkIntegrationTest`?

**Comment by ldaley** [comment]
@oehme I have made those changes. See https://github.com/gradle/gradle/pull/3298/files#diff-0302f6e1670be7ca1b88c87505e0425bR78

**Comment by ldaley** [comment]
@breskeby I just ran `testingBase:check`. Please verify everything else still works.

File: subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestExecutionBuildOperationsIntegrationTest.groovy (hunk 29 above)
Comment (breskeby) [reply subprojects/testing-base/src/integTest/groovy/org/gradle/testing/TestExecutionBuildOperationsIntegrationTest.groovy]: @big-guy I'll address this in a follow on on master. In general we also want XCTests exposed by that.

**Comment by breskeby** [comment]
merged into master after follow up with @ldaley . Some cleanup regarding test coverage will take place on master
